README.md
# siginfo

A module to conveniently get status or debug info from running Python scripts.

<code>siginfo</code> listens for <code>SIGUSR1</code> or <code>SIGUSR2</code>[1] signals and prints info about the current callstack, local variables etc.

## Use cases
Imagine you work line by line on a very large file (Something that happens quite frequenly in Bioinformatics) and you want to see the progress of your program without printing status of each line to <code>STDOUT</code>.<br>
Simply add the following two lines:

```python
from siginfo import SiginfoBasic
SiginfoBasic()
```

Now, run your script and send a <code>SIGUSR1</code> signal to the running process.
```bash
kill -SIGUSR1 ${pid}
```

![image](https://user-images.githubusercontent.com/875703/52584898-a18b9500-2e33-11e9-9d9c-ba2539f3dfb2.png)

*[1]: In addition, on MacOS or BSD system, the <code>SIGINFO</code> signal can be used as well.*

## Installation
Not yet done. Clone the git repo and create a symlink from your python script folder to the <code>signint</code> folder.
```bash
ln -s ./siginfo <PATH_TO_SIGINFO_REPO>/siginfo
```

## Usage
Include <code>siginfo</code> in your python script
```python
import siginfo

# Initialize SignInfo Class
siginfo.SiginfoBasic()
```

### General <code>signinfo</code> classes
<code>siginfo</code> contains the following classes:
- <code>SiginfoBasic</code> Print info about the current stack (and caller stacks). Regular execution continues automatically.
- <code>SigInfoPDB</code> Open the <code>PDB</code> debugger. Pauses script execution until debugger is exited.
- <code>SigInfoSingle</code> Print the value of a single variable of the current scope. Continues regular execution automatically.

#### Initiating the class
All class allow the following arguments:
- <code>info</code> Listen for <code>SIGNFO</code> (Default: <code>True</code>) (only on Mac and BSD)
- <code>usr1</code> Listen for <code>SIGUSR1</code> (Default: <code>True</code>)
- <code>usr2</code> Listen for <code>SIGUSR2</code> (Default: <code>False</code>)
- <code>output</code> Where to write the output to (Default: <code>sys.stdout</code>). Can be anything that offers a <code>write</code> function.

```python
from siginfo import SiginfoBasic
SiginfoBasic(info=True, usr1=False)  # listen only for SIGINFO
SiginfoBasic(info=False, usr=True)  # listen only for SIGUSR1
SiginfoBasic(output=open('mylog.log', 'a'))  # Write call stack output to a log file
```

#### <code>signinfo</code> class instance attributes
- <code>COLUMNS</code>: Maximum width of the Terminal (or max number of rows per line in an output file) (Default: current tty columns - 20; Fallback to 80 if determination isn't possible)
- <code>MAX_LEVELS</code>: Number of stack frames to print (Default: 1 [only the current one])
- <code>OUTPUT</code>: Same as <code>output</code> argument to the constructor function. Defines where to write the output to (Default: <code>sys.stdout</code>)

```python
from siginfo import SiginfoBasic
info_handler = SiginfoBasic()

info_handler.COLUMNS = 200  # Format output to match 200 columns
info_handler.MAX_LEVELS = 4  # Print the current frame + 3 parent frames
info_handler.OUTPUT = open('mylog.log', 'a')  # write the output to mylog.log
```

### API docs
For a more detailed API description, check out https://esbme.com/siginfo/docs/

CHANGELOG.rst
Changelog
=========

0.10
----
- Create CICD for PyPI
- Host documentation on Read The Docs
- Minor fixes, add some types
- Remove `Python 2.7` compatibility

0.9
---
- Updated documentation
- Preparing for pip distribution

tests/test_utils.py
import unittest

from siginfo.utils import left_string


class LeftString(unittest.TestCase):
    def test_trimming(self):
        str_in = '12345'
        str_out = left_string(str_in, 4)
        assert str_out[-3:] == '...'
        assert len(str_out) == 4

    def test_padding(self):
        str_in = '12345'
        str_out = left_string(str_in, 7)
        assert str_out[-3:] == '5  '
        assert len(str_out) == 7

    def test_equal_length(self):
        str_in = '12345'
        str_out = left_string(str_in, 5)
        assert str_out == str_in
        assert len(str_out) == 5


if __name__ == '__main__':
    unittest.main()


tests/test_locals.py
import unittest

from siginfo.localclass import LocalClass


class LocalTest(unittest.TestCase):
    def test_init(self):
        my_locals = {
            'a': 12,
            'b': 'a string',
            'c': {'foo': 'bar'}
        }
        res = LocalClass(my_locals)
        assert isinstance(res, LocalClass), (res, type(res))
        assert len(res.types) == 4
        assert len(res.values) == 4
        assert len(res.var_names) == 4
        for key in my_locals:
            assert type(my_locals[key]).__name__ in res.types
        for idx, key in enumerate(res.var_names):
            if idx == 0:
                # skip the header
                continue
            # Check that the values are identical
            assert str(res.values[idx]) == str(my_locals[key])
            # Check that types are corretly printed
            # We can't use isinstance, we have to compare the strings
            assert res.types[idx] == type(my_locals[key]).__name__  # noqa: E721

    def test_max_width(self):
        loc = LocalClass({'xyz:': 123})
        assert (loc.MAX_KEY+loc.MAX_TYPE+loc.MAX_VALUES+6) <= 80

        loc = LocalClass({'xyz:': 123}, 100)
        assert (loc.MAX_KEY+loc.MAX_TYPE+loc.MAX_VALUES+6) <= 100

        loc = LocalClass({'xyz:': 123}, 20)
        assert (loc.MAX_KEY+loc.MAX_TYPE+loc.MAX_VALUES+6) <= 20

    def test_private_scaling(self):
        loc = LocalClass({'xyz:': 123})
        assert loc.MAX_KEY == 35
        assert loc.MAX_TYPE == 16
        assert loc.MAX_VALUES == 21

        loc._scale_columns(100)
        assert loc.MAX_KEY == 45
        assert loc.MAX_TYPE == 21
        assert loc.MAX_VALUES == 27

        loc._scale_columns(20)
        assert loc.MAX_KEY == 7
        assert loc.MAX_TYPE == 2
        assert loc.MAX_VALUES == 3

    def test_private_add_header(self):
        loc = LocalClass({'xyz:': 123})
        assert loc.var_names[0] == 'VARIABLE'
        assert loc.var_names[1] == 'xyz:'
        assert loc.types[0] == 'TYPE'
        assert loc.types[1] == 'int'
        assert loc.values[0] == 'VALUE'
        assert loc.values[1] == '123'

        loc._add_headers()
        assert loc.var_names[0] == 'VARIABLE'
        assert loc.var_names[1] == 'VARIABLE'
        assert loc.var_names[2] == 'xyz:'
        assert loc.types[0] == 'TYPE'
        assert loc.types[1] == 'TYPE'
        assert loc.types[2] == 'int'
        assert loc.values[0] == 'VALUE'
        assert loc.values[1] == 'VALUE'
        assert loc.values[2] == '123'

    def test_private_make_row(self):
        loc = LocalClass({'xyz:': 123})
        res = loc._make_row('foo', 'bar', 'foobar')
        assert res == 'foo{sp1}| bar{sp2}| foobar{sp3}'.format(
            sp1=' '*33,
            sp2=' '*14,
            sp3=' '*15
        )

    def test_stringify(self):
        loc = LocalClass({'xyz:': 123})
        res = str(loc)
        rows = res.split('\n')
        assert 'VARIABLE' in rows[0]
        assert 'TYPE' in rows[0]
        assert 'VALUE' in rows[0]

        assert 'xyz:' in rows[1]
        assert '| int' in rows[1]
        assert '| 123' in rows[1]


if __name__ == '__main__':
    unittest.main()


tests/test_siginfo.py
import unittest
from siginfo import siginfoclass as si
import sys

OLD_OUT = sys.stdout


class MockOutput(object):
    def __init__(self):
        self.lines = []

    def write(self, line):
        self.lines.append(line)

    def flush(self):
        pass


class MockSignal(object):
    def __init__(self, info=True, usr1=True, usr2=True):
        if info:
            self.SIGINFO = 1
        if usr1:
            self.SIGUSR1 = 2
        if usr2:
            self.SIGUSR2 = 3
        self.signals = []

    def signal(self, sigtype, func):
        self.signals.append((sigtype, func))


class MockClass(object):
    def __init__(self, attrs):
        for key in attrs:
            self.__setattr__(key, attrs[key])
        self._keys = attrs.keys()

    def __str__(self):
        return 'MockClass: {}'.format(', '.join(
            ['{}={}'.format(key, self.__getattribute__(key)) for key in self._keys]
        ))


class MockFrame(object):
    def __init__(self, local_vars={}, line_number=0, back=None):
        self.f_locals = local_vars
        self.f_code = MockClass(
            {'co_name': 'my_test_function_line_{}'.format(line_number)}
        )
        self.f_lineno = line_number
        self.f_back = back


class MockFunction(object):
    def __init__(self):
        self.called = 0
        self.called_with = []

    def __call__(self, *args, **kwargs):
        self.called += 1
        self.called_with.append((args, kwargs))


class SigInfoInitTests(unittest.TestCase):
    def setUp(self):
        si.sys.stdout = MockOutput()
        si.signal = MockSignal()

    def tearDown(self):
        si.sys.stdout = OLD_OUT

    def test_init(self):
        res = si.SiginfoBasic()
        assert isinstance(res, si.SiginfoBasic)
        assert isinstance(res.pid, int)
        assert isinstance(res.MAX_LEVELS, int)
        assert isinstance(res.COLUMNS, int)
        assert isinstance(res.OUTPUT, MockOutput)

    def test_output(self):
        si.sys.stdout.lines = []
        si.SiginfoBasic(info=True, usr1=True, usr2=True)
        assert len(si.sys.stdout.lines) == 6


class SigInfoInputTests(unittest.TestCase):
    def setUp(self):
        si.sys.stdout = MockOutput()

    def tearDown(self):
        si.sys.stdout = OLD_OUT

    def test_inputs(self):
        si.signal = MockSignal()
        res = si.SiginfoBasic(info=True, usr1=True, usr2=True)
        assert res.signals == ['INFO', 'USR1', 'USR2']

        res = si.SiginfoBasic(info=True, usr1=True, usr2=False)
        assert res.signals == ['INFO', 'USR1']

        res = si.SiginfoBasic(info=True, usr1=False, usr2=False)
        assert res.signals == ['INFO']

        res = si.SiginfoBasic(info=True, usr1=False, usr2=True)
        assert res.signals == ['INFO', 'USR2']

        res = si.SiginfoBasic(info=False, usr1=True, usr2=True)
        assert res.signals == ['USR1', 'USR2']

        si.sys.stdout.lines = []
        res = si.SiginfoBasic(info=False, usr1=False, usr2=False)
        assert res.signals == []
        assert si.sys.stdout.lines[0] == 'No signal specified\n'

    def test_inexistent_inputs(self):
        si.sys.stdout.lines = []
        si.signal = MockSignal(info=False)
        res = si.SiginfoBasic(info=True, usr1=True, usr2=True)
        assert res.signals == ['USR1', 'USR2']
        assert 'No SIGINFO availale\n' == si.sys.stdout.lines[0]

    def test_all_missing_inputs(self):
        si.sys.stdout.lines = []
        si.signal = MockSignal(info=False, usr1=False, usr2=False)
        res = si.SiginfoBasic(info=True, usr1=True, usr2=True)
        assert res.signals == []
        assert 'No SIGINFO availale\n' == si.sys.stdout.lines[0]
        assert 'No SIGUSR1 availale\n' == si.sys.stdout.lines[1]
        assert 'No SIGUSR2 availale\n' == si.sys.stdout.lines[2]


class SigInfoSigFormattingTests(unittest.TestCase):
    def test_column_specification(self):
        si.subprocess.check_output = lambda x: '5 140'
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=MockOutput())
        assert res.COLUMNS == 120

        # Defaulting to minimum of 80 colunns
        si.subprocess.check_output = lambda x: '5 79'
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=MockOutput())
        assert res.COLUMNS == 80

        # Defaulting to minimum of 80 colunns
        si.subprocess.check_output = lambda x: '5 81'
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=MockOutput())
        assert res.COLUMNS == 80

        # Defaulting to minimum of 80 colunns
        si.subprocess.check_output = lambda x: '5 101'
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=MockOutput())
        assert res.COLUMNS == 81

        # this will cause the check_output to fail
        si.subprocess.check_output = lambda x: 1
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=MockOutput())
        assert res.COLUMNS == 80


@unittest.skip('Not yet testing script generation')
class SigInfoSigScriptTests(unittest.TestCase):

    def test_create_info_script(self):
        pass

    def test_make_scripts_excecutable(self):
        pass

    def test_script_pids(self):
        pass


class SiginfoFramePrinting(unittest.TestCase):
    def test_print_without_parent(self):
        si.subprocess.check_output = lambda x: '5 80'
        mock_out = MockOutput()
        mock_frame = MockFrame()
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)

        res._print_frame(mock_frame)
        assert len(mock_out.lines) == 16, mock_out.lines
        assert mock_out.lines[1] == 'METHOD\t\tmy_test_function_line_0\n'
        assert mock_out.lines[2] == 'LINE NUMBER:\t0\n'
        assert mock_out.lines[3] == '-'*80
        assert mock_out.lines[5] == 'LOCALS\n'
        assert mock_out.lines[10] == 'SCOPE\t'
        assert mock_out.lines[11] == 'MockClass: co_name=my_test_function_line_0'
        assert mock_out.lines[13] == 'CALLER\t'
        assert mock_out.lines[14] == 'NONE'

    def test_print_with_parent(self):
        si.subprocess.check_output = lambda x: '5 80'
        mock_out = MockOutput()
        mock_frame = MockFrame(back=MockFrame(line_number=2))
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)

        res._print_frame(mock_frame)
        assert len(mock_out.lines) == 16, mock_out.lines
        assert mock_out.lines[1] == 'METHOD\t\tmy_test_function_line_0\n'
        assert mock_out.lines[2] == 'LINE NUMBER:\t0\n'
        assert mock_out.lines[3] == '-'*80
        assert mock_out.lines[5] == 'LOCALS\n'
        assert mock_out.lines[10] == 'SCOPE\t'
        assert mock_out.lines[11] == 'MockClass: co_name=my_test_function_line_0'
        assert mock_out.lines[13] == 'CALLER\t'
        assert mock_out.lines[14] == 'MockClass: co_name=my_test_function_line_2'


class SiginfoCalling(unittest.TestCase):
    def test_signal_calling(self):
        si.subprocess.check_output = lambda x: '5 80'
        mock_out = MockOutput()
        mock_frame = MockFrame()
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)

        res._print_frame = MockFunction()

        res(1, mock_frame)

        assert len(mock_out.lines) == 10
        assert res._print_frame.called == 1
        assert res._print_frame.called_with[0][0][0] == mock_frame

    def test_signal_calling_multiple_level(self):
        si.subprocess.check_output = lambda x: '5 80'
        mock_out = MockOutput()
        mock_frame_back = MockFrame(line_number=2)
        mock_frame = MockFrame(back=mock_frame_back)
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)

        res._print_frame = MockFunction()

        res(1, mock_frame)

        assert len(mock_out.lines) == 16
        assert res._print_frame.called == 2
        assert res._print_frame.called_with[0][0][0] == mock_frame
        assert res._print_frame.called_with[1][0][0] == mock_frame_back

    def test_signal_calling_limit_levels(self):
        """
        2 levels of stack frames are present
        but MAX_LEVELS is set to 1
        """
        si.subprocess.check_output = lambda x: '5 80'
        mock_out = MockOutput()
        mock_frame_back = MockFrame(line_number=2)
        mock_frame = MockFrame(back=mock_frame_back)
        res = si.SiginfoBasic(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)
        res.MAX_LEVELS = 1

        res._print_frame = MockFunction()

        res(1, mock_frame)

        assert len(mock_out.lines) == 10
        assert res._print_frame.called == 1
        assert res._print_frame.called_with[0][0][0] == mock_frame


class SiginfoSingleCalling(unittest.TestCase):
    def test_setting_variables(self):
        mock_out = MockOutput()
        res = si.SigInfoSingle(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)
        res.set_var('foo')

        assert res._varname == 'foo'
        assert res._default is None

    def test_getting_existing_variables(self):
        mock_out = MockOutput()
        mock_frame = MockFrame({'foo': '12', 'bar': 'abc'})
        res = si.SigInfoSingle(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)
        res.set_var('foo')

        mock_out.lines = []
        res(1, mock_frame)

        assert len(mock_out.lines) == 1
        assert mock_out.lines[0] == '12\n'

    def test_getting_nonexisting_variables(self):
        mock_out = MockOutput()
        mock_frame = MockFrame({'foo': '12', 'bar': 'abc'})
        res = si.SigInfoSingle(
            info=False,
            usr1=False,
            usr2=False,
            output=mock_out)
        res.set_var('xyz', 'foobar')

        mock_out.lines = []
        res(1, mock_frame)

        assert len(mock_out.lines) == 1
        assert mock_out.lines[0] == 'foobar\n'


if __name__ == '__main__':
    unittest.main()


siginfo/utils.py
def left_string(string: str, padding: int) -> str:
    """
    Formats a string with padding. Either adds spaces to the end or truncates

    Args
    ----
    string : str
        String for printing
    padding : int
        Width (in columns) for the final format

    Returns
    -------
    : str


    If ``string`` is longer than padding, it will be trimmed with ellipses

    Example
    -------
        ::

            left_string('A very long string', 10)
            # => 'A very ...'

            left_string('short', 10)
            # => 'short     '

    """
    if len(string) > padding:
        return '{}...'.format(string[0:padding-3])
    elif len(string) == padding:
        return string
    else:
        # len(string) < padding
        return '{string}{padding}'.format(
            string=string,
            padding=' '*(int(padding)-len(string))
        )


siginfo/localclass.py
import math

from siginfo.utils import left_string


class LocalClass:
    """
    Formats the ``locale`` object for display in the tty.

    Can handle any kind  of object to format it in a

    +-----+------+-------+
    | KEY | TYPE | VALUE |
    +-----+------+-------+

    like table

    Args
    ----
    local_vars : object
        Object to display in table. key will be one row
    columns : int
        Width (in columns) of the output stream. Default: 80

    """
    def __init__(self, local_vars, columns=80):
        self.var_names = list(local_vars.keys())
        self.types = [type(local_vars[key]).__name__ for key in self.var_names]
        self.values = [str(local_vars[key]) for key in self.var_names]

        self._add_headers()

        self._scale_columns(columns)

    def _scale_columns(self, columns):
        """
        Adjusts the total width of all three columns
        to match the max column width available
        """
        cur_max_key = max(len(val) for val in self.var_names)
        cur_max_type = max(len(val) for val in self.types)
        cur_max_value = max(len(val) for val in self.values)

        cur = cur_max_key + cur_max_type + cur_max_value
        factor = float(columns) / float(cur)
        self.MAX_KEY = math.floor(factor * cur_max_key) - 2
        self.MAX_TYPE = math.floor(factor * cur_max_type) - 2
        self.MAX_VALUES = math.floor(factor * cur_max_value) - 2

    def _add_headers(self):
        self.var_names.insert(0, 'VARIABLE')
        self.types.insert(0, 'TYPE')
        self.values.insert(0, 'VALUE')

    def _make_row(self, key, tp, val):
        return '{key} | {tp} | {val}'.format(
            key=left_string(key, self.MAX_KEY),
            tp=left_string(tp, self.MAX_TYPE),
            val=left_string(val, self.MAX_VALUES)
            )

    def __str__(self):
        res = []
        for idx, key in enumerate(self.var_names):
            res.append(
                self._make_row(key, self.types[idx], self.values[idx])
            )
        return '\n'.join(res)


siginfo/siginfoclass.py
import sys
import signal
import os
import stat
import atexit
import subprocess

from siginfo.localclass import LocalClass


class SiginfoBasic:
    """
    Base class for the SigInfo module

    Args
    ----
    info : bool
        Listen for SIGINFO
        Default: True
        (works only on Mac and BSD)
    usr1 : bool
        Listen for SIGUSR1
        Default: True
    usr2 : bool)
        Listen for SIGUSR2
        Default: True
    output : _io.TextIOWrapper
        IO interface for writing output and log.
        Default: sys.stdout


    Attributes
    ----------
    COLUMNS: int
        Width of Terminal (number of columns)
        Default: Auto (Fallback to 80)
    MAX_LEVELS: int
        Number of parent stack frames to display
        Default: 0 (only current frame)

    Returns
    -------
    SigInfoBasic: An instance of the class

    Example
    -------
        ::

            foo = SiginfoBasic()

            # Write up to 120 characters per line
            foo.COLUMNS = 120

            # Write three parent stack frame
            foo.MAX_LEVELS = 3

            # create executeable to send signal to Python script
            foo.create_info_script('/usr/local/bin')

            # Code of your regular Python script
            def read_lines():
                a = 12
                b = 15
                i = 0
                print('Loading a very long file')
                with open('many_rows.txt') as fh:
                    for line in fh:
                        i += 1
                        # print(line)
                        time.sleep(1)
                print('Done loading')

            # Some useless function so we have more stacks
            def main2():
                read_lines()

            def main():
                main2()

            if __name__ == '__main__':
                main()

        In another terminal window:

        .. code-block:: bash

            # send signal via custom script
            /usr/local/bin/siginfo-USR1

            # Output:
            ========================================================================================================================
            LEVEL       0
            METHOD      read_lines
            LINE NUMBER:    33
            ------------------------------------------------------------------------------------------------------------------------
            LOCALS
            VARIABLE | TYPE            | VALUE
            i        | int             | 1
            fh       | TextIOWrapper   | <_io.TextIOWrapper name='many_rows.txt' mode='r' encoding='UTF-8'>
            b        | int             | 15
            a        | int             | 12
            line     | str             | Row 1

            ------------------------------------------------------------------------------------------------------------------------
            SCOPE   <code object read_lines at 0x108c30c90, file "long_script.py", line 24>
            CALLER  <code object main2 at 0x108c309c0, file "long_script.py", line 21>
            ========================================================================================================================

            ========================================================================================================================
            LEVEL       1
            METHOD      main2
            LINE NUMBER:    22
            ------------------------------------------------------------------------------------------------------------------------
            LOCALS
            VARIABLE                                               | TYPE                       | VALUE
            ------------------------------------------------------------------------------------------------------------------------
            SCOPE   <code object main2 at 0x108c309c0, file "long_script.py", line 21>
            CALLER  <code object main at 0x108c30ed0, file "long_script.py", line 18>
            ========================================================================================================================

            ...


    """
    def __init__(self, info=True, usr1=True, usr2=False, output=None):
        self.COLUMNS = 80
        self.MAX_LEVELS = 0  # How many parent stack frames to display
        self.OUTPUT = output or sys.stdout  # where to print the output to
        self.pid = os.getpid()
        self.signals = []

        # Bind SIGINFO if available and requested
        if info:
            if hasattr(signal, 'SIGINFO'):
                signal.signal(signal.SIGINFO, self)
                self.OUTPUT.write('Listening for >>SIGINFO<<\n')
                self.OUTPUT.write('==> kill -s INFO {}\n'.format(self.pid))
                self.signals.append('INFO')
            else:
                self.OUTPUT.write('No SIGINFO availale\n')

        # Bind SIGUSR1 if available and requested
        if usr1:
            if hasattr(signal, 'SIGUSR1'):
                signal.signal(signal.SIGUSR1, self)
                self.OUTPUT.write('Listening for >>SIGUSR1<<\n')
                self.OUTPUT.write('==> kill -s USR1 {}\n'.format(self.pid))
                self.signals.append('USR1')
            else:
                self.OUTPUT.write('No SIGUSR1 availale\n')

        # Bind SIGUSR2 if available and requested
        if usr2:
            if hasattr(signal, 'SIGUSR2'):
                signal.signal(signal.SIGUSR2, self)
                self.OUTPUT.write('Listening for >>SIGUSR2<<\n')
                self.OUTPUT.write('==> kill -s USR2 {}\n'.format(self.pid))
                self.signals.append('USR2')
            else:
                self.OUTPUT.write('No SIGUSR2 availale\n')

        if not info and not usr1 and not usr2:
            self.OUTPUT.write('No signal specified\n')
        self.OUTPUT.flush()

        # Attempts to use all columns of the current tty window size
        # Falls back to 80 columns by default
        try:
            rows, columns = subprocess.check_output(['stty', 'size']).split()
            self.COLUMNS = max([self.COLUMNS, int(columns)-20])
        except Exception:
            self.COLUMNS = 80

    def create_info_script(self, path=None, prefix='', overwrite=False):
        """
        Create an executable on the file system to send the appropiate signal.

        For user convenience, create executable filess in the specified path
        that can be used to send corresponding signals to the parent's script.

        Args
        ----
        path : str
            path to executable files (default $HOME)
        prefix : str
            Prefix of the executable file name (default '')
        overwrite : bool
            Overwrite existing executables (default False)

        Returns
        ------
        None

        """
        if path is None:
            path = os.path.expanduser('~')
        for sig in self.signals:
            filename = os.path.abspath(
                os.path.join(
                    path,
                    '{}siginfo-{}'.format(prefix, sig)
                )
            )
            if not os.path.isfile(filename) or overwrite:
                with open(filename, 'w') as fh:
                    fh.write('#!/bin/sh\n')
                    fh.write('kill -s {} {}'.format(sig, self.pid))
                os.chmod(filename, os.stat(filename).st_mode | stat.S_IEXEC)

                atexit.register(self._delete_file, filename)

    def _print_frame(self, frame):
        """
        Formats and prints the frame output
        in a somewhat tabbular format
        """
        local_vars = LocalClass(frame.f_locals, self.COLUMNS)
        self.OUTPUT.write('METHOD\t\t{}\n'.format(frame.f_code.co_name))
        self.OUTPUT.write('LINE NUMBER:\t{}\n'.format(frame.f_lineno))
        self.OUTPUT.write('-'*self.COLUMNS)
        self.OUTPUT.write('\n')
        self.OUTPUT.write('LOCALS\n')
        self.OUTPUT.write(str(local_vars))
        self.OUTPUT.write('\n')
        self.OUTPUT.write('-'*self.COLUMNS)
        self.OUTPUT.write('\n')
        self.OUTPUT.write('SCOPE\t')
        self.OUTPUT.write(str(frame.f_code))
        self.OUTPUT.write('\n')
        self.OUTPUT.write('CALLER\t')
        if frame.f_back:
            self.OUTPUT.write(str(frame.f_back.f_code))
        else:
            self.OUTPUT.write('NONE')
        self.OUTPUT.write('\n')

    # Print all stack frames
    # callback for signal.signal
    def _call(self, signum, frame):
        depth = self.MAX_LEVELS or 1000
        self.OUTPUT.write('\n')
        self.OUTPUT.write(type(self).__name__)
        self.OUTPUT.write('\n')

        for i in range(depth):
            if frame:
                self.OUTPUT.write('\n')
                self.OUTPUT.write('='*self.COLUMNS)
                self.OUTPUT.write('\n')
                self.OUTPUT.write('LEVEL    \t{}\n'.format(i))
                self._print_frame(frame)
                self.OUTPUT.write('='*self.COLUMNS)
                self.OUTPUT.write('\n')
                self.OUTPUT.flush()
                frame = frame.f_back
            else:
                self.OUTPUT.flush()
                break

    __call__ = _call

    @staticmethod
    def _delete_file(filename):
        """
        used for atexit cleanup
        """
        if os.path.isfile(filename):
            os.remove(filename)


class SigInfoPDB(SiginfoBasic):
    """
    SigInfo class that starts the Python PDB Debugger

    Instead of printing out the call stack, it allows to
    interactively inspect the current frame via PDB Debugger.

    Local variables from current frame are avaialable via the  ``_locals`` dict.

    Example
    -------
        ::

            foo = SigInfoPDB()
            do_long_task()

            # create executeable to send signal to Python script
            foo.create_info_script('/usr/local/bin')

        Now get access to PDB debugger in another terminal window:

        .. code-block:: bash

            # send signal via custom script
            /usr/local/bin/siginfo-USR1

    """

    # Start the PDB Debugger
    def __call__(self, signum, frame):
        self._call(signum, frame)
        _locals = frame.f_locals
        print('Waiting for your command')
        import pdb; pdb.Pdb(nosigint=True).set_trace()


class SigInfoSingle(SiginfoBasic):
    """
    SigInfo class that only returns a single value

    Example
    -------
        ::

            # create signal listener
            foo = SigInfoSingle(usr1=True)

            # define which variable to print
            foo.set_var('i')

            # create executeable to send signal to Python script
            foo.create_info_script('/usr/local/bin')

            # run your regular code
            for n in seq(1, 100):
                do_long_task(i=n)

        In another terminal window:

        .. code-block:: bash

            # send signal via custom script
            /usr/local/bin/siginfo-USR1

            # Output:
            4


    """
    def set_var(self, varname, default=None):
        """
        Defines the variable that should be printed

        Args
        ----
        varname : str
            Name of variable from stack frame that should be printed
        default
            Backup value to print if ``varname`` is not present in local
            stack frame.
            Default: None

        Returns
        -------
        None
        """
        self._varname = varname
        self._default = default

    # Print value of set variable
    def __call__(self, signum, frame):
        if self._varname:
            self.OUTPUT.write('{}\n'.format(
                frame.f_locals.get(self._varname, self._default)
            ))
